File: 6_tree/树状数组/307. 区域和检索 - 数组可修改.ts
import { BIT } from './树状数组单点更新模板'

/**
 * 动态区间和求解
 */
class NumArray {
  private nums: number[]
  private bit: BIT

  constructor(nums: number[]) {
    this.nums = nums
    this.bit = new BIT(nums.length)
    for (let i = 0; i < this.bit.size; i++) {
      this.bit.add(i + 1, nums[i])
    }
  }

  update(index: number, val: number): void {
    this.bit.add(index + 1, val - this.nums[index])
    this.nums[index] = val
  }

  sumRange(left: number, right: number): number {
    return this.bit.query(right + 1) - this.bit.query(left)
  }
}

const numArray = new NumArray([1, 3, 5])

console.log(numArray.sumRange(0, 2))
numArray.update(1, 2)
console.log(numArray.sumRange(0, 2))

File: 6_tree/线段树/307. 区域和检索 - 数组可修改.py
from typing import List
from python线段树模板 import SegmentTree


class NumArray:
    def __init__(self, nums: List[int]):
        self.nums = nums
        self.tree = SegmentTree(nums)

    def update(self, index: int, val: int) -> None:
        self.tree.update(index + 1, index + 1, val - self.nums[index])
        self.nums[index] = val

    def sumRange(self, left: int, right: int) -> int:
        return self.tree.query(left + 1, right + 1)


# Your NumArray object will be instantiated and called as such:
obj = NumArray([1, 3, 5])
print(obj.sumRange(0, 2))
obj.update(1, 2)
print(obj.sumRange(0, 2))

# 9 8
obj = NumArray([-1, 2, 3])
print(obj.sumRange(0, 2))
obj.update(0, 1)
print(obj.sumRange(0, 0))
